File: cloud-api/cloud-api-v2-client.py
#!/usr/bin/python3

import requests
import time
import json
import threading
import logging 
import os 
import queue
import argparse


NUM_THREADS = 30  # Number of threads
TASKS_PER_AUTHENTICATION = 50 # Number of tasks before re-authentication

SLEEP_TIME = 10.0 # seconds wait time between querying request
MAX_CHECK_STATUS = 1000 # Retry 1000 times to check status before stopping
USERNAME: str
PASSWORD: str

def parser() -> argparse.ArgumentParser:
    parser = argparse.ArgumentParser(
            description="Celantur Cloud API v2 Client",
            formatter_class=argparse.ArgumentDefaultsHelpFormatter)
    parser.add_argument("-i", "--input", help="Input folder containing images", required=True)
    parser.add_argument("-o", "--output", help="Output folder for saving anonymized images", required=True)
    parser.add_argument("-u", "--username", help="Username for Celantur Cloud API", required=True)
    parser.add_argument("-p", "--password", help="Password for Celantur Cloud API", required=True)
    parser.add_argument("-c", "--configuration", help="Anonymisation configuration as JSON file", required=True)
    parser.add_argument("-e", "--endpoint", help="Celantur Cloud API v2 endpoint", default='https://api.celantur.com/v2/')
    return parser



def setup_logger(log_dir='logs'):
    # Create the log directory if it doesn't exist
    os.makedirs(log_dir, exist_ok=True)

    # Get the current timestamp
    timestamp = time.strftime('%Y-%m-%d_%H-%M-%S')

    # Define the log file name with timestamp
    log_file = f'{log_dir}/celantur-api-v2-client_{timestamp}.log'

    # Create the logger and set its level to DEBUG
    logger = logging.getLogger('my_logger')
    logger.setLevel(logging.DEBUG)

    # Create the file handler
    file_handler = logging.FileHandler(log_file)

    # Create the console handler
    console_handler = logging.StreamHandler()

    # Define the log format
    log_format = '[%(asctime)s] [%(threadName)s] [%(levelname)s]: %(message)s'
    date_format = '%Y-%m-%d %H:%M:%S'
    formatter = logging.Formatter(log_format, date_format)

    # Set the formatter for both handlers
    file_handler.setFormatter(formatter)
    console_handler.setFormatter(formatter)

    # Add the handlers to the logger
    logger.addHandler(file_handler)
    logger.addHandler(console_handler)

    return logger

logger = setup_logger()


def get_files_from_(root_path: str, relative_path: str = "") -> str:
    path = os.path.join(root_path, relative_path)
    extensions = ['.jpg', '.jpeg', '.png']
    for file in os.scandir(path):
        if os.path.isfile(file) and os.path.splitext(file)[1].lower() in extensions:
            yield os.path.join(relative_path, file.name)
        elif os.path.isdir(file):
            yield from get_files_from_(root_path, os.path.join(relative_path, file.name))


def get_files_without_overwrite_from_(root_input_path: str, root_output_path: str, input_queue: queue.Queue):
    """
    Put input file paths into the queue
    """
    for file_path in get_files_from_(root_input_path):
        output_path = os.path.join(root_output_path, file_path)
        if os.path.exists(output_path):
            logger.info(f"Skip {file_path} because {output_path} already exists.")
        else:
           input_queue.put((root_input_path, file_path))
           logger.debug(f"Put into file queue: {file_path}")
           


def authenticate():
  data = {'username': USERNAME, 'password': PASSWORD}
  response = requests.post(ENDPOINT_LOGIN, json=data, headers={'Content-Type':'application/json'})
  resp_dict = response.json()
  try:
    auth_token = resp_dict['AccessToken']
    # TODO implement logic for token renewal before expiration
    # expires_in = resp_dict['ExpiresIn']  # 3600s
    logger.info('🔒 Successfully authenticated and token received.')
    return auth_token
  except:
     logger.error(f'Login error (Status {response.status_code}): {response.text}')
     raise SystemExit(-1)
     


def load_image(file_path: str):
  image = ''
  try:
    image_file = open(file_path,'rb')
    image = image_file.read()
    image_file.close()
    logger.debug(f'🖼️ Image loaded: {file_path}.')
  except Exception as e:
    logger.info(f'Could not read image: {e}')
  return image


def create_task(anonymisation_configuration: str, auth_token: str):
  response = requests.post(ENDPOINT_TASK, data=json.dumps(anonymisation_configuration), headers={'Authorization': auth_token})

  if response.status_code == 200:
    response_body = response.json()
    
    task_id = response_body['task_id']
    upload_url = response_body['upload_url']

    logger.info(f"Task {task_id} created.")
    return (task_id, upload_url)
  else:
    logger.error(f'Creating task failed (Status {response.status_code}): {response.text}')


def get_task_status(task_id: str, auth_token: str):
  # stati: new, queued, processing, done or failed
  status_url = f'{ENDPOINT_TASK}{task_id}/status'

  response = requests.get(status_url, headers={'Authorization': auth_token})  
  if response.status_code == 200:
    response_body = response.json()
    status = response_body['task_status']
    logger.info(f'Task {task_id} has status: {status}')
    return status
  else:
    logger.error(f'Getting task status failed (Status {response.status_code}): {response.text}')
    return response.status_code
  

def get_task(task_id: str, auth_token: str):
  # stati: new, queued, processing, done or failed
  task_url = f'{ENDPOINT_TASK}{task_id}'

  response = requests.get(task_url, headers={'Authorization': auth_token})  
  response_body = response.json()
  if response.status_code == 200:
    logger.info(f'GET /task/{task_id} successful.')
    return response_body
  else:
    logger.error(f'Getting task faild (Status {response.status_code}): {response.text}')
    return 0


def upload_image(file_path: str, upload_url: str):
  img = load_image(file_path)
  response = requests.put(url=upload_url, data=img)
  if response.status_code == 200:
    logger.info(f'⬆️ Uploaded image {file_path} successfully.')
    return True
  else:
    logger.error(f'Image upload failed (Status {response.status_code}): {response.text}')
    return False


def download_image(output_file_name: str, task_id: str, auth_token: str, sleep_time: float):
  global total_count
  counter = 1
  while counter < MAX_CHECK_STATUS:
    task_status = get_task_status(task_id, auth_token)
    if task_status == "done":
      break
    logger.info(f"[Retry {counter}/{MAX_CHECK_STATUS}] Status: {task_status}, sleeping {sleep_time} seconds ...")
    counter += 1
    time.sleep(sleep_time)
  else:
     if task_status != 'done':
        logger.warning(f"The task {task_id} did not finish.")
  
  task = get_task(task_id, auth_token)

  anonymized_url = task['anonymized_url']
  response = requests.get(anonymized_url)
  os.makedirs(os.path.dirname(output_file_name), exist_ok=True)  
  with open(output_file_name, 'wb') as f:
    f.write(response.content)
  logger.info(f'[image {total_count}] Anonymized image {output_file_name} received 👍.')
  logger.info(f'Task {task_id} completed.')

def run_test(input_queue: queue.Queue, output_folder: str, anonymisation_configuration: dict):
  global auth_token, total_count
  while not input_queue.empty():
    input_root_path, relative_file_path = input_queue.get()
    task_id, upload_url = create_task(anonymisation_configuration, auth_token)
    input_file_path = os.path.join(input_root_path, relative_file_path)
    output_file_path = os.path.join(output_folder, relative_file_path)
    if upload_image(input_file_path, upload_url):
      download_image(output_file_path, task_id, auth_token, SLEEP_TIME)
    total_count += 1
    if total_count % TASKS_PER_AUTHENTICATION == 0:  # Re-authenticate
        auth_token = authenticate()


def read_configuration_file(file_name: str) -> dict:
   if file_name is None:
      raise ValueError("No configuration specified!")
   else:
    with open(file_name, 'r') as fp:
       return json.load(fp)


def create_thread(*run_test_args) -> threading.Thread:
  thread = threading.Thread(target=run_test, args=run_test_args)
  thread.start()
  return thread


if __name__ == "__main__":
    total_count = 0
    # Measure the execution time
    start_time = time.time()
    args = parser().parse_args()
    endpoint: str = args.endpoint.rstrip('/')
    configuration: dict = read_configuration_file(args.configuration)

    ENDPOINT_LOGIN = f'{endpoint}/signin/'
    ENDPOINT_TASK = f'{endpoint}/task/'
    USERNAME = args.username
    PASSWORD = args.password
    

    logger.info(f"Start Cloud API v2 Client with {NUM_THREADS} threads.")

    input_queue = queue.Queue(maxsize=100)
    file_reader = threading.Thread(name="ReadInput", target=get_files_without_overwrite_from_, args=(args.input, args.output, input_queue))
    file_reader.start()

    auth_token = authenticate()

    threads = [create_thread(input_queue, args.output, configuration) for _ in range(NUM_THREADS)]
    for t in threads:
        t.join()
    file_reader.join()
    end_time = time.time()

    # Calculate the elapsed time
    elapsed_time = end_time - start_time
    logger.info(f"Completed. It took {int(elapsed_time)} seconds ({elapsed_time//60} minutes)")

File: cloud-api/cloud-api-v2-client.js
import Queue from 'yocto-queue';
import path from 'path';
import fs from 'fs';
import { pipeline } from 'stream/promises';

const TASKS_PER_AUTHENTICATION = 50; // Number of tasks before re-authentication
const SLEEP_TIME = 10000; // ms wait time between querying request
const MAX_CHECK_STATUS = 1000; // Retry 1000 times to check status before stopping
const EXTENSIONS = ['.jpg', '.jpeg', '.png'];

export class CelanturClientParameters {
    constructor(
        input, 
        output, 
        username, 
        password, 
        configuration, 
        endpoint,  
        recursive,
        fileType) {

        this.input = input;
        this.output = output;
        this.username = username;
        this.password = password;
        this.configuration = configuration;
        this.endpoint = endpoint;
        this.recursive = recursive;
        this.fileType = fileType;
    }

}

export class CelanturCloudAPIClient {
    // constructor fields
    #inputQueue = new Queue(); // queue of filepaths to process
    #authToken;
    #totalCount = 0; 
    #inputFolder;
    #outputFolder;
    #username;
    #password;
    #configFilePath;
    #endpoint;
    #recursive;
    #fileType;
    
    // init. fields on run()
    #config;
    #dottedExtensions;
    #ENDPOINT_LOGIN;
    #ENDPOINT_TASK;


    constructor(params) {
        if (!params.input || !params.output || !params.username || !params.password || !params.configuration) {
            throw Error("Missing required arguments: input, output, username, password or configuration.");
        }
        this.#inputFolder = params.input;
        this.#outputFolder = params.output;
        this.#username = params.username;
        this.#password = params.password;
        this.#configFilePath = params.configuration;
        this.#endpoint = params.endpoint !== undefined ? params.endpoint : 'https://api.celantur.com/v2/';
        this.#recursive = params.recursive !== undefined ? params.recursive : true;
        this.#fileType = params.fileType;
    }
    
    async run() {
        // setup based on args
        this.#dottedExtensions = this.normalizeExtensions(this.#fileType);
        this.#config = await this.readConfigurationFile();

        this.#endpoint = this.#endpoint.replace(/\/$/, ''); //remove trailing slash (if any)
        this.#ENDPOINT_LOGIN = `${this.#endpoint}/signin/`;
        this.#ENDPOINT_TASK = `${this.#endpoint}/task/`;

        await this.getFilesWithoutOverwriteFrom(this.#inputFolder, this.#outputFolder, this.#recursive, this.#dottedExtensions);

        this.#authToken = await this.authenticate();
        const startTime = Date.now();
        const processingTasks = [];
        while (this.#inputQueue.size !== 0) {
            processingTasks.push(this.runTask(this.#inputQueue.dequeue()));
        }

        Promise.all(processingTasks).then(() => {
            const endTime = Date.now();
            console.info(
                `Completed. It took ${Math.floor((endTime - startTime) / 1000)} seconds (${Math.floor((endTime - startTime) / 60000)} minutes)`
            );
        });
    }

    async runTask(taskFile) {
        const taskData = await this.createTask();
        taskData.input_file_path = path.join(taskFile.rootInputPath, taskFile.relFilePath);
        taskData.output_file_path = path.join(this.#outputFolder, taskFile.relFilePath);
        
        if (await this.uploadImage(taskData)) {
            await this.downloadImage(taskData);
            this.#totalCount++;
    
            if (this.#totalCount % TASKS_PER_AUTHENTICATION == 0) this.#authToken = await this.authenticate(); // re-authenticate
        }
    
    }

    async loadImage(path) {
        try {
          const buffer = await fs.promises.readFile(path);
          console.log(`Image loaded: ${path}`);
          return buffer;    
        } catch (err) {
          console.error('Error reading image: ', err);
          throw err;
        }
    }
    
    async uploadImage(task) {
        const img = await this.loadImage(task.input_file_path);
        
        const response = await fetch(task.upload_url, {
            method: 'PUT',
            body: img
        });
        if (!response.ok) {
            throw new Error(`Image upload failed (Status ${response.status_code}): ${response.text}`);
        }
    
        console.log(`Uploaded image ${task.input_file_path} successfully.`);
        return true;
      }
    
    
    async downloadImage(task) {
        let counter = 1;
        let taskStatus;
        while (counter < MAX_CHECK_STATUS) {
            taskStatus = await this.getTaskStatus(task.task_id);
            if (taskStatus === 'done') {
                break;
            }
            console.log(`[Retry ${counter}/${MAX_CHECK_STATUS}] Status: ${taskStatus}, sleeping ${SLEEP_TIME} ms...`);
            counter++;
            await new Promise(r => setTimeout(r, SLEEP_TIME));
        }
    
        if (taskStatus != 'done') {
            console.log(`The task ${task.task_id} did not finish in time.`)
        } else {
            const taskResponse = await this.getTask(task.task_id);
            const anonymizedUrl = taskResponse.anonymized_url;
            const response = await fetch(anonymizedUrl);
            
            const dir = path.dirname(task.output_file_path);
            try { // check if dir exists (if not, create)
                await fs.promises.access(dir);
            } catch (err) {
                if (err.code === 'ENOENT') {
                    await fs.promises.mkdir(dir, { recursive: true });
                } else {
                    throw err;
                }
            }
            await pipeline(response.body, fs.createWriteStream(task.output_file_path));
    
            console.info(`[image ${this.#totalCount}] Anonymized image ${task.output_file_path} received.`);
            console.info(`Task ${task.task_id} completed.`);
        }
    }
    
    
    async getTaskStatus(id) {
        const response = await fetch(`${this.#ENDPOINT_TASK}${id}/status`, {
        method: 'GET',
        headers: { 'Authorization': this.#authToken}
        });
        
        if (!response.ok) {
            throw new Error(`Getting task status failed (Status ${response.status_code}): ${response.text}`);
        }
        const responseBody = await response.json(); 
        console.log(`Task ${id} has status: ${responseBody.task_status}`);
        return responseBody.task_status;
    }
    
    async getTask(id) {
        const response = await fetch(`${this.#ENDPOINT_TASK}${id}`, {
        method: 'GET',
        headers: {'Authorization': this.#authToken}
        });
        
        if (!response.ok) {
            throw new Error(`Getting task failed (Status ${response.status_code}): ${response.text}`);
        }
        console.info(`GET /task/${id} successful.`);
        const responseBody = await response.json(); 
        return responseBody;
    }
    
    async readConfigurationFile() {
        try {
            const data = await fs.promises.readFile(this.#configFilePath, 'utf8');
            return JSON.parse(data);
        } catch (err) {
            console.error('Error reading the file:', err);
        }
    }
    
    // Ensure that the file extensions start with a dot and are lowercase.
    normalizeExtensions(extensions) {
        if (!extensions || extensions.length === 0) {
            return EXTENSIONS;
        }
        else {
            let dottedExtensions = [];
            for (let e of extensions) {
                let lowercase = e.toLowerCase()
                if (!lowercase.startsWith('.')) {
                    lowercase = '.' + lowercase;
                }
                if (!EXTENSIONS.includes(lowercase)) {
                    throw new Error(`File type ${lowercase} not supported!`);
                } else {
                    dottedExtensions.push(lowercase);
                }
            }
            return dottedExtensions;
        }
    }
    
    async *getFilesFrom(rootPath, relativePath, extensions, recursive) {
        const dirPath = path.join(rootPath, relativePath);
    
        try {
            const files = await fs.promises.readdir(dirPath);
            
            for (const file of files) {
                const filePath = path.join(dirPath, file);
                const stat = await fs.promises.stat(filePath);
                
                if (stat.isFile() && extensions.includes(path.extname(file).toLowerCase())) {
                    yield path.join(relativePath, file);
                } else if (recursive && stat.isDirectory()) {
                    yield* this.getFilesFrom(rootPath, path.join(relativePath, file), extensions, recursive);
                }
            }
        } catch (err) {
            console.error(err);
            process.exit(-1);
        }
    }
    
    async getFilesWithoutOverwriteFrom(rootInputPath, rootOutputPath, recursive, extensions) {
        for await (const relFilePath of this.getFilesFrom(rootInputPath, '', extensions, recursive)) {
            const outputPath = path.join(rootOutputPath, relFilePath);
            
            try {
                const outputExists = await fs.promises.access(outputPath).then(() => true).catch(() => false);
                
                if (outputExists) {
                    console.log(`Skip ${relFilePath} because ${outputPath} already exists.`);
                } else {
                    this.#inputQueue.enqueue({rootInputPath: rootInputPath, 
                        relFilePath: relFilePath});
                        console.log(`Put into file queue: ${relFilePath}`);
                    }
                } catch (err) {
                console.error(`Error checking file: ${err.message}`);
            }
        }
    }
    
    async authenticate() {
        const data = {
            username: this.#username,
            password: this.#password
        };
    
        try {
            const response = await fetch(this.#ENDPOINT_LOGIN, {
                method: 'POST',
                headers: { 'Content-Type': 'application/json' },
                body: JSON.stringify(data)
            });
            if (!response.ok) {
                throw new Error(`Login error (Status ${response.status}): ${await response.text()}`);
            }
    
            const responseBody = await response.json();
            const authToken = responseBody.AccessToken;
    
            console.info('Successfully authenticated and token received.');
            return authToken;
        } catch (err) {
            console.error(err.message);
            process.exit(-1);
        }
    }
    
    async createTask() {
        const response = await fetch(this.#ENDPOINT_TASK, {
            method: 'POST',
            headers: { 'Authorization': this.#authToken },
            body: JSON.stringify(this.#config)
        });
        if (!response.ok) {
            throw new Error(`Creating task failed (Status ${response.status}): ${await response.text()}`);
        }
    
        const responseBody = await response.json();
        console.log(`Task ${responseBody.task_id} created`);
        return responseBody;
    }
}